docs/arch-split.md

```diff
@@ -254,12 +271,40 @@ In a generic theory, we typically only interpret the Arch locale:
 - to keep existing proofs checking until we find time to factor out the
   architecture-dependent parts.
 
+
+### Unconventional requalification shortcut
+
+While the expected convention is to perform requalify commands in the generic
+theory as described above, there exists a shortcut for doing so in
+architecture-specific theories when outside the Arch context:
+
+```isabelle
+requalify_facts
+  ARM.user_mem_dom_cong
+
+thm user_mem_dom_cong      (* ok *)
+thm ARM.user_mem_dom_cong  (* ok *)
+thm Arch.user_mem_dom_cong (* ERROR *)
+```
+
+This immediately makes the fact available in the global context. While it is a
+violation of expected conventions and needs to be repeated in every
+arch-specific theory file, there is one important difference:
+* the `.` in `context begin interpretation Arch .` in the middle of AInvs takes 7.5s
+* `requalify_facts` in the global context is nearly instant (even for
+multiple facts).
+
+This disparity will only get worse as the Arch context grows bigger, and
+might indicate the need for some alternative functionality.
+(FIXME: check if this is true)
```

Comment:
So, it is true that it does get worse: the more lemmas there are in the locale, the longer it takes to enter. There have been some improvements there from upstream, but not dramatically so. However: we already are using the Arch locale in the current unsplit Refine, and splitting should actually reduce the number of lemmas that are inside the locale. On the other hand, we will be entering it more often. So it may go up, or go down, but I don't think it will change dramatically.

Reply:
Well even if it stays the same, it's still ~7s per invocation vs essentially free. Would it make any sense to find some kind of way to shortcut it, e.g. something that would understand L4V_ARCH and maybe be able to add `_H` or `_A`, so it would be (just throwing out ideas): `requalify_facts_arch foo -> requalify_facts ARM.foo`? Does it seem at all worth it? Would also get rid of the duplication of the current ones that live towards the end of Arch theories.

Reply:
That does seem like a worthwhile idea and can probably be implemented on top of the current `requalify` commands. We could provide the `_A` and `_H` postfixes as an optional parameter if we wanted to. They would be consistent across architectures, they only indicate which abstraction level of the proof the name is from. Let's turn this into an issue and investigate further.

Reply:
Cool. I'll make an issue.

Reply:
https://github.com/seL4/l4v/issues/777